File: HashTable/387. First Unique Character in a String.py
class Solution:
    def firstUniqChar(self, s: str) -> int:
        dict = {} #hash map

        for i,n in enumerate(s):
            if n not in dict:
                dict[n] = i
            else:
                dict[n] = -1

        smallest = 100001
        #remove the duplicates with comprehension?

        dictComp = {key: value for key, value in dict.items() if value != -1}
        if len(dictComp) == 0:
            return -1

        for index in dictComp.values():
            if index < smallest:
                smallest = index
        return smallest

File: HashTable/387. First Unique Character in a String.java
class Solution {
    public int firstUniqChar(String s) {
        //initialize a hash map
        Map<Character, Integer> hashmap = new HashMap<>();
        //char[] charArray = s.toCharArray();

        for (int index = 0; index < s.length(); index++) {
            //fill up and call duplicates -1
            if(!hashmap.containsKey(s.charAt(index))){
                hashmap.put(s.charAt(index), index);
            }
            else{
                hashmap.put(s.charAt(index), 100001);
            }
        }
        int smallVal = 100001;
        /*
        //remove duplicates
        for (Map.Entry<Character, Integer> entry : hashmap.entrySet()){
            if(entry.getValue() == -1){
                hashmap.remove(entry);
            }
        } */
        
        if(hashmap.size() == 0) return -1;

        //find the smallest
        for (Map.Entry<Character, Integer> entry : hashmap.entrySet()){
            if(entry.getValue() < smallVal){
                smallVal = entry.getValue();
            }
        }
        if(smallVal == 100001) return -1;
        return smallVal;
        
    
        
    }
}